File: app/src/main/java/i/am/shiro/amai/AmaiApplication.java
package i.am.shiro.amai;

import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.GlideBuilder;

import io.realm.Realm;
import io.realm.RealmConfiguration;
import timber.log.Timber;

import static i.am.shiro.amai.constant.Constants.DEFAULT_CHANNEL_ID;

public class AmaiApplication extends Application {

    @Override
    public void onCreate() {
        super.onCreate();

        Preferences.init(this);

        if (BuildConfig.DEBUG) initDebugTools();

        Realm.init(this);
        RealmConfiguration config = new RealmConfiguration.Builder()
                .schemaVersion(1)
                .migration(new Migration())
                .compactOnLaunch()
                .build();
        Realm.setDefaultConfiguration(config);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel mChannel = new NotificationChannel(
                    DEFAULT_CHANNEL_ID,
                    getString(R.string.app_name),
                    NotificationManager.IMPORTANCE_DEFAULT);

            NotificationManager notificationManager = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
            if (notificationManager != null) {
                notificationManager.createNotificationChannel(mChannel);
            }
        }
    }

    private void initDebugTools() {
        Timber.plant(new Timber.DebugTree() {
            @Override
            protected String createStackElementTag(@NonNull StackTraceElement element) {
                String tag = super.createStackElementTag(element);
                String method = element.getMethodName();
                return String.format("%s:%s", tag, method);
            }
        });

        GlideBuilder glideBuilder = new GlideBuilder()
            .setLogLevel(Log.VERBOSE);
        Glide.init(this, glideBuilder);
    }
}

File: app/src/main/java/i/am/shiro/amai/AmaiApplication.kt
package i.am.shiro.amai

import android.app.Application
import coil3.ImageLoader
import coil3.PlatformContext
import coil3.SingletonImageLoader
import coil3.network.okhttp.OkHttpNetworkFetcherFactory
import coil3.util.DebugLogger
import i.am.shiro.amai.coil3.PageInterceptor
import i.am.shiro.amai.coil3.ThumbnailInterceptor
import i.am.shiro.amai.koin.mainModule
import okhttp3.OkHttpClient
import org.koin.android.ext.android.get
import org.koin.android.ext.koin.androidContext
import org.koin.core.context.startKoin
import timber.log.Timber

class AmaiApplication : Application(), SingletonImageLoader.Factory {

    override fun onCreate() {
        super.onCreate()

        if (BuildConfig.DEBUG) initDebugTools()

        startKoin {
            androidContext(this@AmaiApplication)
            modules(mainModule)
        }
    }

    private fun initDebugTools() {
        Timber.plant(object : Timber.DebugTree() {
            override fun createStackElementTag(element: StackTraceElement): String {
                val tag = super.createStackElementTag(element)
                val method = element.methodName
                return "$tag:$method"
            }
        })
    }

    override fun newImageLoader(context: PlatformContext) = ImageLoader.Builder(context)
        .components {
            add(ThumbnailInterceptor())
            add(PageInterceptor())
            add(OkHttpNetworkFetcherFactory(get<OkHttpClient>()))
        }
        .logger(DebugLogger())
        .build()
}